rust/connlib/tunnel/src/device_channel/tun_windows.rs

```diff
@@ -20,24 +23,40 @@ use windows::Win32::{
     },
     Networking::WinSock::{AF_INET, AF_INET6},
 };
+use windows::Win32::{
+    Foundation::{BOOLEAN, HANDLE},
+    System::Threading::{
+        RegisterWaitForSingleObject, UnregisterWaitEx, INFINITE, WT_EXECUTEINWAITTHREAD,
+    },
+};
 
 // TODO: Double-check that all these get dropped gracefully on disconnect
 pub struct Tun {
     _adapter: Arc<wintun::Adapter>,
     /// The index of our network adapter, we can use this when asking Windows to add / remove routes / DNS rules
     /// It's stable across app restarts and I'm assuming across system reboots too.
     iface_idx: u32,
-    // TODO: Get rid of this mutex. It's a hack to deal with `poll_read` taking a `&self` instead of `&mut self`
-    packet_rx: std::sync::Mutex<mpsc::Receiver<wintun::Packet>>,
-    _recv_thread: std::thread::JoinHandle<()>,
+    // Arc because the recv calls are implemented on Arc
     session: Arc<wintun::Session>,
+    wait_handle: Win32::Foundation::HANDLE,
+    // Pin<Box> needed so that we can pass a pointer to the wait handle callback
+    // and it probably won't move. (I might be mis-understanding Pin here)
+    // Mutex needed because AtomicWaker has a debug assert if you call `register` concurrently
+    waker: Mutex<std::pin::Pin<Box<AtomicWaker>>>,
```

Comment:
You don't need the `Mutex` here, the waker is already atomic, it's enough to just wrap the `AtomicWaker` in an `Arc`, that is guaranteed to not move or drop, as long as you don't decrement the reference count.

Reply:
No it was panicking here in debug mode https://docs.rs/futures-core/0.3.30/src/futures_core/task/__internal/atomic_waker.rs.html#355
Adding the mutex seemed to fix that.
Are there multiple threads calling `poll_read` at once?

I could go back to the Arc if I clone it every time we register the waker, since otherwise the callbacks messed up the refcount and freed it, causing memory safety problems.

Reply:
> Are there multiple threads calling poll_read at once?
There shouldn't be but also that shouldn't cause a panic.

> I could go back to the Arc if I clone it every time we register the waker, since otherwise the callbacks messed up the refcount and freed it, causing memory safety problems.
well, yeah, you need to increase the refcount when sending it to the callback otherwise you might drop while there's still a reference.

Reply:
I think it panicked because there is that `debug_assert` in AtomicWaker's code if you try to register from two threads at once. I can try to get debug logs to prove that.

I can go back to Arc and store the handle of the in-flight callback I guess.
I think when I call `UnregisterEx` on the wait handle, it will not call the callback, so the `Arc<AtomicWaker>` that was cloned and sent to the callback will leak. That is not unsafe but it's also just one more thing to keep in mind